spec/ddtrace/sampler_spec.rb

```diff
@@ -0,0 +1,177 @@
+
+  describe '#sample' do
+    let(:spans) do
+      [
+        Datadog::Span.new(nil, '', trace_id: 1),
+        Datadog::Span.new(nil, '', trace_id: 2),
+        Datadog::Span.new(nil, '', trace_id: 3)
+      ]
+    end
+
+    it 'samples all spans' do
+      spans.each do |span|
+        expect(sampler.sample(span)).to be true
+        expect(span.sampled).to be true
+      end
+    end
+  end
+end
+
+RSpec.describe Datadog::RateSampler do
+  subject(:sampler) { described_class.new(sample_rate) }
+
+  before(:each) { Datadog::Tracer.log.level = Logger::FATAL }
+  after(:each) { Datadog::Tracer.log.level = Logger::WARN }
+
+  describe '#initialize' do
+    context 'given a sample rate' do
+      context 'that is negative' do
+        let(:sample_rate) { -1.0 }
+        it { expect(sampler.sample_rate).to eq(1.0) }
+      end
+
+      context 'that is 0' do
+        let(:sample_rate) { 0.0 }
+        it { expect(sampler.sample_rate).to eq(1.0) }
+      end
+
+      context 'that is between 0 and 1.0' do
+        let(:sample_rate) { 0.5 }
+        it { expect(sampler.sample_rate).to eq(0.5) }
+      end
+
+      context 'that is greater than 1.0' do
+        let(:sample_rate) { 1.5 }
+        it { expect(sampler.sample_rate).to eq(1.0) }
+      end
+    end
+  end
+
+  describe '#sample' do
+    let(:spans) do
+      [
+        Datadog::Span.new(nil, '', trace_id: 1),
+        Datadog::Span.new(nil, '', trace_id: 2),
+        Datadog::Span.new(nil, '', trace_id: 3)
+      ]
+    end
+
+    shared_examples_for 'rate sampling' do
+      let(:span_count) { 1000 }
+      let(:rng) { Random.new(123) }
+
+      let(:spans) { Array.new(span_count) { Datadog::Span.new(nil, '', trace_id: rng.rand(Datadog::Span::MAX_ID)) } }
+      let(:expected_num_of_sampled_spans) { span_count * sample_rate }
+
+      it 'samples an appropriate proportion of spans' do
+        spans.each do |span|
+          sampled = sampler.sample(span)
+          expect(span.get_metric(Datadog::RateSampler::SAMPLE_RATE_METRIC_KEY)).to eq(sample_rate) if sampled
+        end
+
+        expect(spans.select(&:sampled).length).to be_within(expected_num_of_sampled_spans * 0.1)
+          .of(expected_num_of_sampled_spans)
+      end
+    end
+
+    it_behaves_like('rate sampling') { let(:sample_rate) { 0.1 } }
+    it_behaves_like('rate sampling') { let(:sample_rate) { 0.25 } }
+    it_behaves_like('rate sampling') { let(:sample_rate) { 0.5 } }
+    it_behaves_like('rate sampling') { let(:sample_rate) { 0.9 } }
+
+    context 'when a sample rate of 1.0 is set' do
+      let(:sample_rate) { 1.0 }
+
+      it 'samples all spans' do
+        spans.each do |span|
+          expect(sampler.sample(span)).to be true
+          expect(span.sampled).to be true
+          expect(span.get_metric(Datadog::RateSampler::SAMPLE_RATE_METRIC_KEY)).to eq(sample_rate)
+        end
+      end
+    end
+  end
+end
+
+RSpec.describe Datadog::PrioritySampler do
```

Comment:
Let's add coverage to check that, when there is no client sampling, `SAMPLE_RATE_METRIC_KEY` isn't set.

Reply:
Added this.